proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+   ``f`` or ``g`` without declaring these as arguments to ``h`` and passing
+   them explicitly. Fixing this was the subject of `#40`_; see `Future Work`_.
+
+4. If a function or group of functions needs to work with a datatype or class
+   locally, there is no way to do this without polluting the namespace of the
+   entire file.
+
+   With this proposal, we can do this easily::
+
+     module (f) where
+       f :: ...
+       f = ...
+
+       data SpecialDataType = ...
+
+       class LocalClass a b c where ...
+
+5. When we expect users to import our library qualified, we have no way of
+   signaling this beyond documentation; we also have no way of making it easy
+   to import the module qualified correctly.
+
+   With this proposal, we can do this easily::
+
+     -- top of file:
+     module Data.Set ( Set, module qualified Set ) where
+
+       module Set ( Set, fromList ) where
+         data Set = ...
+         fromList = ...
+
+   An importer who days ``import Data.Set`` will get access to ``Set`` (the
+   type) and ``Set.fromList``, the function. The fact that the module and type
+   have the same name is inconsequential here, but it is permitted.
+
+6. Suppose our library exports class ``C`` with method ``meth``. Our users
+   will frequently import ``C(..)`` and get ``meth`` in scope. Now I wish to
+   refactor ``C`` without changing my users' import behavior. There is no way
+   to do this currently.
+
+   With this proposal, we can do this::
+
+     -- top of file:
+     module MyLibrary ( module C(meth) ) where
+
+       class C a where ...
+
+       module C where
+         meth :: ...
+
+   This example shows that modules may be *extended*. The ``class C``
+   declaration implicitly creates module ``C``, which is then extended below.
+
+   An import specifier of ``C(..)`` will not import ``meth`` after this
+   change. Instead, importers must say ``module C`` in the import list.
+   If they want unqualified access to ``C``\ 's contents, they will additionally
+   have to write ``import module C`` separately. This new form works to replace
+   today's ``C(..)`` and is customizable, and so may become the recommended way
+   of importing a type with its auxiliary definitions.
+
+7. When importing a library qualified, there are often stretches of code which
+   work with the imported type, performing many operations on values of that type.
+   In these stretches, the qualifications can become noisy. ::
+
+     module Main where
+
+     import qualified Data.Set as Set
+
+     frobbleSets :: Set.Set Bool -> Set.Set Int -> Set.Set Char
+     frobbleSets = ...
+
+   In ``frobbleSets``, there will be many ``Set.`` qualifications.
+
+   With this proposal, we could add ::
+
+     where
+       import module Set
+
+   to the end of the definition of ``frobbleSet`` to bring all the exports
+   from ``Data.Set`` into scope unqualified, locally. This reduces noise and
+   may serve to encourage users to import more identifiers qualified.
```

Comment:
The `where import module Set` will not help with the noisy `Set.` qualifiers in `frobbleSets`’s type signature. Not surprising, I was just briefly confused because the example above has `Set.` qualifiers _only_ in the type signature, whereas this local import will only help with the `Set.` qualifiers hidden in `...`.